src/nvme/core.c

```diff
@@ -755,3 +757,72 @@ void nvme_close(struct nvme_ctrl *ctrl)
 
 	memset(ctrl, 0x0, sizeof(*ctrl));
 }
+
+int nvme_configure_cmb(struct nvme_ctrl *ctrl)
+{
+	uint64_t cmbmsc;
+	uint32_t cmbloc;
+	uint32_t cmbsz;
+	uint64_t cap;
+	int bar;
+
+	cap = le64_to_cpu(mmio_read64(ctrl->regs + NVME_REG_CAP));
+	if (!NVME_FIELD_GET(cap, CAP_CMBS))
+		return 0;
+
+	cmbmsc = NVME_FIELD_SET(1, CMBMSC_CRE);
+	mmio_lh_write64(ctrl->regs + NVME_REG_CMBMSC, cpu_to_le64(cmbmsc));
+
+	cmbloc = le32_to_cpu(mmio_read32(ctrl->regs + NVME_REG_CMBLOC));
+	bar = NVME_FIELD_GET(cmbloc, CMBLOC_BIR);
+
+	cmbmsc = le64_to_cpu(mmio_read64(ctrl->regs + NVME_REG_CMBMSC));
+	cmbsz = le32_to_cpu(mmio_read32(ctrl->regs + NVME_REG_CMBSZ));
+
+	ctrl->cmb.bar = bar;
+	ctrl->cmb.size = nvme_cmb_size(cmbsz);
+	ctrl->cmb.vaddr = vfio_pci_map_bar(&ctrl->pci, bar,
+			ctrl->cmb.size, NVME_FIELD_GET(cmbloc, CMBLOC_OFST),
+			PROT_READ | PROT_WRITE);
+	if (!ctrl->cmb.vaddr) {
+		log_debug("could not map bar %d vaddr\n", bar);
+		return -1;
+	}
+
+	if (iommu_map_vaddr(__iommu_ctx(ctrl), ctrl->cmb.vaddr,
+				ctrl->cmb.size, &ctrl->cmb.iova, 0x0)) {
+		log_debug("could not map bar vaddr to iommu\n");
+		return -1;
+	}
+
+	cmbmsc |= NVME_FIELD_SET(1, CMBMSC_CMSE) |
+		(ctrl->cmb.iova >> NVME_CMBMSC_CBA_SHIFT) <<
+		NVME_CMBMSC_CBA_SHIFT;
+
+	mmio_lh_write64(ctrl->regs + NVME_REG_CMBMSC, cpu_to_le64(cmbmsc));
```

Comment:
I think this needs to be a `mmio_hl_write64` (we want to make sure we write the CBA *before* setting CMSE).

Reply:
Ah, my bad, actually I intended just like you're saying, but reversered ;) Thanks for catching this!

Reply:
This is still the low-high variant.

Reply:
Missed that one.  Applied, Thanks.